### PR TITLE
Add pr-context-synthesis skill

### PR DESCRIPTION
**Pull request by AryanGodara**
# What this is

A primitive for producing a tight 1–3 paragraph what / why / how synthesis of a single PR (or PR-shaped change). Stays anchored to the actual diff and refuses vague verbs.

Ships with `/pr-synthesis <N|owner/repo#N|url>`, a slash command that fetches PR metadata, linked issue, and diff and prints the synthesis verbatim.

# Why split out

Extracted from #4351 because the synthesis primitive is reusable beyond a single review report — review and incident-investigation flows both want the same per-PR summary shape. Splitting keeps the contract small.

# Files

- `docs/skills/pr-context-synthesis.md` — rules + output shape.
- `.claude/commands/pr-synthesis.md` — slash command wrapping the procedure.

#### Example 1: PR #4243

Boundary case: just over the gate. Generated bindings dominate the line count; the substantive change is a 275-line estimator + a 236-line e2e test.

Adds a new `Eip4626` native price estimator to the price-estimation stack at `crates/price-estimation/src/native/eip4626.rs`. The estimator unwraps an EIP-4626 vault token by calling `asset()` and `convertToAssets(10^decimals)` on-chain, then delegates pricing of the underlying token to the next estimator in the configured stage. New contract bindings (`IERC4626`, `MockERC4626Wrapper`) ship as pre-generated alloy crates; an e2e forked-mainnet test covers a single vault and a recursive vault chain.

Vault tokens like sDAI often lack direct DEX liquidity, which makes the existing native-price estimators fall over. This change lets the orderbook quote vault tokens by reducing them to their underlying. (No linked issue; intent inferred from the PR body and the new test names.)

Mechanically: a `Mutex<HashSet<Address>>` negative cache short-circuits subsequent requests for tokens whose `asset()` reverts (cleared on restart); each on-chain RPC is bounded by `tokio::time::timeout`, with the leftover budget forwarded to the inner estimator so total wall-clock stays inside the caller's timeout — important for nested vaults. A configurable `depth` parameter (default 1) controls recursion, wired in `factory.rs::create_native_estimator`. Config deserialisation rejects stages where `Eip4626` is the last entry, since it must be followed by another estimator to price the underlying.

*Diff summarised at file-scope only — total +4,411 / −21 across 24 files exceeds the 2k full-diff threshold. Generated bindings (`ierc4626/src/lib.rs` +648, `mockerc4626wrapper/src/lib.rs` +2,768) and the `MockERC4626Wrapper.json` artifact (+240) account for ~83% of additions; the substantive Rust changes are the 275-line estimator, 236-line e2e test, and 14 small wiring edits.*

#### Example 2: PR #4217

Stress case: two orders of magnitude over the gate. Demonstrates that the per-file bucket summary is genuinely **more useful than the full diff would be** — 96% of the lines are mechanical codegen output, and the substantive change hides in a 116-file `+213/−1,531` workspace bucket of import-path edits.

Moves the `contracts` crate from `crates/contracts` to a standalone `contracts/` directory at repo root, and replaces its `build.rs` runtime code-generation with 148 pre-generated per-contract Alloy binding crates under `contracts/generated/contracts-generated/`. A new `contracts-facade` crate re-exports the bindings so consumers see a single import surface. All 116 workspace crates that previously depended on `crates/contracts` are rewired to import from the new generated crates. Codegen still exists, but is now a separate `cargo run -p contracts` step rather than an implicit cost on every workspace build.

Driver: `build.rs` was the longest critical-path step in the workspace build. The PR description's flame graphs claim ~15 s shaved off cold builds (from ~60 s to ~45 s) and visibly higher CPU utilisation since the build no longer serialises behind a single codegen blocker. Pre-generated bindings are also independently version-controllable and reviewable — historically the only way to inspect a binding diff was to rerun the build.

*Diff summarised at file-scope only — +375,175 / −1,654 across 384 files is two orders of magnitude over the 2k full-diff threshold. Bucketed: generated bindings 148 files / +292,618 (mechanical, alloy-codegen output); contract artifacts 74 files / +7,278 (JSON ABI moves); workspace crates 116 files / +213 / −1,531 (overwhelmingly import-path edits and removal of the old `crates/contracts/`); Cargo.lock 3 files / +7,747; contracts tooling 38 files / +1,192. Almost everything except the 116 workspace-crate imports is mechanically generated or moved.*

## Reviewing this in isolation

The skill body is self-contained. The intro mentions consumers (`/review-pr`, ad-hoc summaries, future incident walks) but doesn't link to anything that doesn't exist on `main`. Rules, Shape, and Example stand on their own.

# How to try it

```
/pr-synthesis 4371
```

### CODE REVIEW

File: .claude/commands/pr-synthesis.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+---
+description: Produce a tight 1–3 paragraph what / why / how synthesis of a single PR. Fetches title, body, linked issue, and diff scope; outputs per `docs/skills/pr-context-synthesis.md`. Useful as a standalone summary or as a building block for review / incident-investigation flows. Read-only.
```
Comment (jmg-duarte) [inline .claude/commands/pr-synthesis.md]: I don't think this bit is necessary
```suggestion
description: Produce a tight 1–3 paragraph what / why / how synthesis of a single PR. Fetches title, body, linked issue, and diff scope; outputs per `docs/skills/pr-context-synthesis.md`. Read-only.
```

File: .claude/commands/pr-synthesis.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+---
+
+Synthesise PR: $ARGUMENTS
+
+## Parse $ARGUMENTS
+
+Accept any of:
+
+- A PR number: `4267`
```
Comment (jmg-duarte) [inline .claude/commands/pr-synthesis.md]:
```suggestion
- A PR number: `4267`/`#4267`
```

File: .claude/commands/pr-synthesis.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+and abort.
+
+## Procedure
+
+Fetch in parallel:
+
+```bash
+gh pr view <N> -R <owner>/<repo> --json title,body,files,labels,baseRefName,headRefName,closingIssuesReferences
+gh pr diff <N> -R <owner>/<repo>
```
Comment (jmg-duarte) [inline .claude/commands/pr-synthesis.md]: This diff can potentially be huge and break the context window.
This should first get a summary of the diff size and then decide whether to chunk it or not.

File: .claude/commands/pr-synthesis.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+```
+
+For each entry in `closingIssuesReferences` (GitHub's own parsing of `Fixes #N` / `Closes #N` / `Resolves #N`, more reliable than regexing the body), fetch the issue:
+
+```bash
+gh issue view <issue.number> -R <issue.repository.owner>/<issue.repository.name> --json title,body,labels,state
+```
+
+If `closingIssuesReferences` is empty, proceed without a linked issue — never invent one.
```
Comment (jmg-duarte) [inline .claude/commands/pr-synthesis.md]: This should honestly be done even before the diff, it should help setup context for the diff and also it would allow Claude to high mismatches

File: docs/skills/pr-context-synthesis.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# Skill — PR context synthesis
+
+Use to produce a tight 1–3 paragraph *what / why / how* block for a single PR (or PR-shaped change). Consumed by the PR review report's CONTEXT section and by [`pr-blame-walk`](pr-blame-walk.md), which calls this skill once per candidate PR before scoring.
+
+## How to invoke
+
+Two ways:
+
+- **Procedurally** — follow the rules below when you need a tight 1–3 paragraph synthesis (CONTEXT block of `/review-pr`, per-candidate block in `pr-blame-walk`, ad-hoc *"summarise this PR for me"*).
```
Comment (jmg-duarte) [inline docs/skills/pr-context-synthesis.md]: Doesn't this mean that this PR is blocked by the one for pr-blame-walk?

File: docs/skills/pr-context-synthesis.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+- **Via slash command** — `/pr-synthesis <N|owner/repo#N|url>` fetches the PR, linked issue, and diff for you and prints the synthesis verbatim.
+
+## Inputs
+
+- `<pr_text>` — PR title and body. If there is no PR yet (e.g. local-diff mode), substitute the branch name plus the relevant commit messages.
```
Comment (jmg-duarte) [inline docs/skills/pr-context-synthesis.md]: A bit more accurate (?)
```suggestion
- `<pr_text>` — PR title and body. If there is no PR yet (e.g. local-diff mode), use the current branch name and the relevant commit messages (i.e. diff against `main`).
```

File: docs/skills/pr-context-synthesis.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+- `<linked_issue>` — title and body of any issue referenced via `Fixes #N` / `Closes #N` / `Resolves #N`. May be empty.
+- `<diff_summary>` — file scope plus a codemap or per-file note of the actual change. The ground truth the synthesis must stay anchored to.
```
Comment (jmg-duarte) [inline docs/skills/pr-context-synthesis.md]: If this is the ground truth, should it be first in the list so the follow up items are interpreted with this in context?

**Review by jmg-duarte — COMMENTED** [review]
Could you share examples in the description?

I also suggest you try these:
* https://github.com/cowprotocol/services/pull/4243
* https://github.com/cowprotocol/services/pull/4217

File: docs/skills/pr-context-synthesis.md (hunk 5 above)
Comment (AryanGodara) [reply docs/skills/pr-context-synthesis.md]: good catch 👀 i dropped the reference, so the skill is self invoked

File: .claude/commands/pr-synthesis.md (hunk 3 above)
Comment (AryanGodara) [reply .claude/commands/pr-synthesis.md]: added a 2K LOC check, (additions + removals), then it falls back to per file changes

File: .claude/commands/pr-synthesis.md (hunk 4 above)
Comment (AryanGodara) [reply .claude/commands/pr-synthesis.md]: true that
reordered: metadata -> issues -> diff

File: docs/skills/pr-context-synthesis.md (hunk 7 above)
Comment (AryanGodara) [reply docs/skills/pr-context-synthesis.md]: yep, that makes sense. reordered this to go first

**Comment by jmg-duarte** [comment]
I retried it on the huge contracts PR and this was so annoying to use, it was constantly asking for `awk` permissions

Let's add some priority rules when reviewing so it doesn't take generated code into account right out of the gate and only if needed, that should help quite a bit
